// common-accessible-components/lib/components/atoms/Button/Button.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _utils = require('../../../utils');

var _reactProptypeConditionalRequire = require('react-proptype-conditional-require');

var _reactProptypeConditionalRequire2 = _interopRequireDefault(_reactProptypeConditionalRequire);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Spinner = require('../Spinner');

var _Spinner2 = _interopRequireDefault(_Spinner);

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _config = require('../../../config');

var _decorators = require('../../../decorators');

var buttonTypes = ['button', 'submit', 'reset'];

var kinds = ['default', 'primary', 'secondary', 'danger', 'link'];

var buttonSizes = ['small', 'large', 'fullWidth'];

var buttonEventNames = ['onClick', 'onTouchStart', 'onTouchEnd', 'onTouchCancel', 'onMouseDown', 'onMouseEnter', 'onMouseLeave', 'onFocus', 'onBlur'];

var Button = (function (_Component) {
    _inherits(Button, _Component);

    function Button() {
        _classCallCheck(this, _Button);

        _get(Object.getPrototypeOf(_Button.prototype), 'constructor', this).apply(this, arguments);
    }

    _createClass(Button, [{
        key: 'render',
        value: function render() {
            var _props = this.props;
            var kind = _props.kind;
            var spinnerClassName = _props.spinnerClassName;
            var children = _props.children;
            var className = _props.className;
            var isLoading = _props.isLoading;
            var disabled = _props.disabled;

            var other = _objectWithoutProperties(_props, ['kind', 'spinnerClassName', 'children', 'className', 'isLoading', 'disabled']);

            return _react2['default'].createElement(
                'button',
                _extends({
                    className: Button.getClasses(this.props),
                    disabled: this.props.disabled || this.props.isLoading
                }, other, Button.getEventsBinding(this)),
                this.props.isLoading ? _react2['default'].createElement(_Spinner2['default'], { className: this.props.spinnerClassName }) : this.props.children
            );
        }
    }], [{
        key: 'getClasses',
        value: function getClasses(props) {
            var _extends2;

            var block = _config.prefix + '-' + Button.displayName.toLowerCase();
            var styleMod = props.kind && block + '--' + props.kind;
            var sizeMod = props.size && block + '--' + props.size;
            var iconMod = props.size && block + '--icon';
            var disabledMod = block + '--disabled';
            var loadingMod = block + '--loading';
            var extraClasses = props.className.split(' ');

            return (0, _classnames2['default'])(_extends((_extends2 = {}, _defineProperty(_extends2, block, true), _defineProperty(_extends2, styleMod, !!styleMod), _defineProperty(_extends2, sizeMod, !!sizeMod), _defineProperty(_extends2, iconMod, Button.isIconButton(props)), _defineProperty(_extends2, disabledMod, props.disabled), _defineProperty(_extends2, loadingMod, props.isLoading), _extends2), Button.getUIClasses(props), _lodash2['default'].fromPairs(_lodash2['default'].map(extraClasses, function (className) {
                return [className, true];
            }))));
        }
    }, {
        key: 'getEventsBinding',
        value: function getEventsBinding(context) {
            return _lodash2['default'].fromPairs(_lodash2['default'].map(buttonEventNames, function (eventName) {
                return [eventName, context.props[eventName].bind(context)];
            }));
        }
    }, {
        key: 'displayName',
        value: 'Button',
        enumerable: true
    }, {
        key: 'propTypes',
        value: _extends({
            'aria-label': (0, _reactProptypeConditionalRequire2['default'])(_react.PropTypes.string, function (props) {
                return Button.isIconButton(props);
            }),
            framework: _react.PropTypes.string.isRequired,
            kind: _react.PropTypes.oneOf(kinds),
            children: _react.PropTypes.any.isRequired,
            className: _react.PropTypes.string,
            disabled: _react.PropTypes.bool,
            spinnerClassName: _react.PropTypes.string,
            isLoading: _react.PropTypes.bool,
            name: _react.PropTypes.string.isRequired,
            size: _react.PropTypes.oneOf(buttonSizes),
            type: _react.PropTypes.oneOf(buttonTypes).isRequired
        }, _lodash2['default'].fromPairs(_lodash2['default'].map(buttonEventNames, function (eventName) {
            return [eventName, _react.PropTypes.func];
        }))),
        enumerable: true
    }, {
        key: 'defaultProps',
        value: _extends({
            className: '',
            disabled: false,
            isLoading: false,
            type: _lodash2['default'].first(buttonTypes)
        }, _lodash2['default'].fromPairs(_lodash2['default'].map(buttonEventNames, function (eventName) {
            return [eventName, _lodash2['default'].noop];
        })), {
            framework: _lodash2['default'].head(Object.keys(_config.uiFrameworks))
        }),
        enumerable: true
    }, {
        key: 'isIconButton',
        value: function value(props) {
            return (0, _utils.hasChild)(props.children, _Icon2['default']) && _react2['default'].Children.count(props.children) === 1;
        },
        enumerable: true
    }]);

    var _Button = Button;
    Button = (0, _decorators.uiFrameworkClasses)(Button) || Button;
    return Button;
})(_react.Component);

exports['default'] = Button;
module.exports = exports['default'];
