// redbox-noreact/lib/index.js
'use strict';

exports.__esModule = true;

var _typeofReactElement = typeof Symbol === 'function' && Symbol['for'] && Symbol['for']('react.element') || 60103;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _errorStackParser = require('error-stack-parser');

var _errorStackParser2 = _interopRequireDefault(_errorStackParser);

var _style = require('./style');

var Redbox = function Redbox(_ref) {
  var error = _ref.error;
  return {
    $$typeof: _typeofReactElement,
    type: 'div',
    key: null,
    ref: null,
    props: {
      children: [{
        $$typeof: _typeofReactElement,
        type: 'div',
        key: null,
        ref: null,
        props: {
          children: [error.name, ': ', error.message],
          style: _style.message
        },
        _owner: null
      }, {
        $$typeof: _typeofReactElement,
        type: 'div',
        key: null,
        ref: null,
        props: {
          children: _errorStackParser2['default'].parse(error).map(function (f, index) {
            var link = f.fileName + ':' + f.lineNumber + ':' + f.columnNumber;
            return {
              $$typeof: _typeofReactElement,
              type: 'div',
              key: index,
              ref: null,
              props: {
                children: [{
                  $$typeof: _typeofReactElement,
                  type: 'div',
                  key: null,
                  ref: null,
                  props: {
                    children: f.functionName
                  },
                  _owner: null
                }, {
                  $$typeof: _typeofReactElement,
                  type: 'div',
                  key: null,
                  ref: null,
                  props: {
                    children: {
                      $$typeof: _typeofReactElement,
                      type: 'a',
                      key: null,
                      ref: null,
                      props: {
                        children: link,
                        href: link,
                        style: _style.linkToFile
                      },
                      _owner: null
                    },
                    style: _style.file
                  },
                  _owner: null
                }],
                style: _style.frame
              },
              _owner: null
            };
          }),
          style: _style.stack
        },
        _owner: null
      }],
      style: _style.redbox
    },
    _owner: null
  };
};

Redbox.displayName = 'RedBox';

exports['default'] = Redbox;
module.exports = exports['default'];
