// common-accessible-components/lib/ui-themes/bootstrap-theme.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});
exports['default'] = {
    row: {
        'default': 'row'
    },
    column: {
        large: function large(x) {
            return 'col-lg-' + x;
        },
        medium: function medium(x) {
            return 'col-md-' + x;
        },
        small: function small(x) {
            return 'col-xs-' + x;
        },
        offsetLarge: function offsetLarge(x) {
            return 'col-lg-offset-' + x;
        },
        offsetMedium: function offsetMedium(x) {
            return 'col-md-offset-' + x;
        },
        offsetSmall: function offsetSmall(x) {
            return 'col-xs-offset-' + x;
        },
        pushLarge: function pushLarge(x) {
            return 'col-lg-push-' + x;
        },
        pushMedium: function pushMedium(x) {
            return 'col-md-push-' + x;
        },
        pushSmall: function pushSmall(x) {
            return 'col-xs-push-' + x;
        }
    },
    container: {
        'default': 'container',
        fluid: 'container-fluid'
    },
    button: {
        'default': 'btn',
        kind: function kind(x) {
            return 'btn-' + x;
        },
        size: function size(x) {
            var mapping = {
                large: 'btn-lg',
                small: 'btn-sm',
                fullWidth: 'btn-block'
            };
            return x && mapping[x];
        }
    },

    image: {
        responsive: 'img-responsive',
        rounded: 'img-rounded',
        circle: 'img-circle',
        thumbnail: 'img-thumbnail'
    },

    input: {
        type: function type(x) {
            return x === 'checkbox' ? '' : 'form-control';
        }
    },

    alert: {
        'default': 'alert',
        type: function type(x) {
            if (x === 'error') {
                return 'alert-danger';
            }

            return 'alert-' + x;
        },
        dismissable: 'alert-dismissable'
    },

    screenReader: {
        text: 'sr-only'
    },

    formGroup: {
        'default': 'form-group',
        invalid: 'has-error'
    },

    spinner: {
        'default': 'fa fa-spinner fa-spin'
    }
};
module.exports = exports['default'];
