// common-accessible-components/lib/components/atoms/Link/Link.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactProptypeConditionalRequire = require('react-proptype-conditional-require');

var _reactProptypeConditionalRequire2 = _interopRequireDefault(_reactProptypeConditionalRequire);

var _ScreenReader = require('../ScreenReader');

var _ScreenReader2 = _interopRequireDefault(_ScreenReader);

var _config = require('../../../config');

var _decorators = require('../../../decorators');

var targets = ['_self', '_blank', '_parent', '_top'];

var linkEventNames = ['onClick', 'onTouchStart', 'onTouchEnd', 'onTouchCancel', 'onMouseDown', 'onMouseEnter', 'onMouseLeave'];

var Link = (function (_Component) {
    _inherits(Link, _Component);

    function Link() {
        _classCallCheck(this, _Link);

        _get(Object.getPrototypeOf(_Link.prototype), 'constructor', this).apply(this, arguments);
    }

    _createClass(Link, [{
        key: 'render',
        value: function render() {
            var _props = this.props;
            var className = _props.className;
            var title = _props.title;
            var srNewWindowText = _props.srNewWindowText;

            var other = _objectWithoutProperties(_props, ['className', 'title', 'srNewWindowText']);

            return _react2['default'].createElement(
                'a',
                _extends({
                    className: Link.getClasses(this.props),
                    title: Link.getTitle(this.props)
                }, other, Link.getEventsBinding(this)),
                this.props.children,
                this.props.target === '_blank' ? _react2['default'].createElement(_ScreenReader2['default'], { text: this.props.srNewWindowText }) : null
            );
        }
    }], [{
        key: 'displayName',
        value: 'Link',
        enumerable: true
    }, {
        key: 'propTypes',
        value: {
            children: _react.PropTypes.any.isRequired,
            href: _react.PropTypes.string.isRequired,
            srNewWindowText: (0, _reactProptypeConditionalRequire2['default'])(_react.PropTypes.string, function (props) {
                return props.target === '_blank';
            }),
            className: _react.PropTypes.string,
            title: _react.PropTypes.string,
            target: _react.PropTypes.oneOf(targets),
            framework: _react.PropTypes.oneOf(Object.keys(_config.uiFrameworks)).isRequired
        },
        enumerable: true
    }, {
        key: 'defaultProps',
        value: _extends({
            className: '',
            framework: _lodash2['default'].head(Object.keys(_config.uiFrameworks))
        }, _lodash2['default'].fromPairs(_lodash2['default'].map(linkEventNames, function (eventName) {
            return [eventName, _lodash2['default'].noop];
        }))),
        enumerable: true
    }, {
        key: 'isTextLink',
        value: function value(_ref) {
            var children = _ref.children;

            var count = _react2['default'].Children.count(children);
            var firstChild = _lodash2['default'].first(_react2['default'].Children.toArray(children));
            return count === 1 && typeof firstChild === 'string';
        },
        enumerable: true
    }, {
        key: 'isOpenInNewWindow',
        value: function value(props) {
            return props.target === '_blank';
        },
        enumerable: true
    }, {
        key: 'getTitle',
        value: function value(props) {
            if (!Link.isOpenInNewWindow(props)) {
                return props.title;
            }
            var linkText = Link.isTextLink(props) && _lodash2['default'].first(_react2['default'].Children.toArray(props.children));
            var title = props.title || linkText || '';

            return title + ' ' + props.srNewWindowText;
        },
        enumerable: true
    }, {
        key: 'getClasses',
        value: function value(props) {
            var block = _config.prefix + '-' + Link.displayName.toLowerCase();
            var extraClasses = props.className && props.className.split(' ');

            return (0, _classnames2['default'])(_extends(_defineProperty({}, block, true), Link.getUIClasses(props), _lodash2['default'].fromPairs(_lodash2['default'].map(extraClasses, function (className) {
                return [className, true];
            }))));
        },
        enumerable: true
    }, {
        key: 'getEventsBinding',
        value: function value(context) {
            return _lodash2['default'].fromPairs(_lodash2['default'].map(linkEventNames, function (eventName) {
                return [eventName, context.props[eventName].bind(context)];
            }));
        },
        enumerable: true
    }]);

    var _Link = Link;
    Link = (0, _decorators.uiFrameworkClasses)(Link) || Link;
    return Link;
})(_react.Component);

exports['default'] = Link;
module.exports = exports['default'];
