// common-accessible-components/lib/components/atoms/HelpMessage/HelpMessage.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Alert = require('../Alert');

var _Alert2 = _interopRequireDefault(_Alert);

var _config = require('../../../config');

var _decorators = require('../../../decorators');

var HelpMessage = function HelpMessage(props) {
    var children = props.children;

    var other = _objectWithoutProperties(props, ['children']);

    return _react2['default'].createElement(
        _Alert2['default'],
        _extends({
            type: 'warning',
            screenreaderText: 'Help:'
        }, other),
        props.children
    );
};

HelpMessage.displayName = 'HelpMessage';

HelpMessage.propTypes = {
    className: _react.PropTypes.string,
    children: _react.PropTypes.any.isRequired,
    framework: _react.PropTypes.oneOf(Object.keys(_config.uiFrameworks))
};

exports['default'] = HelpMessage;
module.exports = exports['default'];
