// nwb/lib/commands/build-react-app.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _webpackBuild = require('../webpackBuild');

var _webpackBuild2 = _interopRequireDefault(_webpackBuild);

// Use a config function, as this won't be called until after NODE_ENV has been
// set by webpackBuild() and we don't want these optimisations in development
// builds.
var buildConfig = function buildConfig() {
  var config = {
    devtool: 'source-map',
    entry: {
      app: _path2['default'].resolve('src/index.js')
    },
    output: {
      filename: '[name].js',
      path: _path2['default'].resolve('public/build'),
      publicPath: '/build/'
    },
    plugins: {
      vendorChunkName: 'vendor'
    }
  };
  if (process.env.NODE_ENV === 'production') {
    config.loaders = {
      babel: {
        query: {
          optional: ['optimisation.react.inlineElements', 'optimisation.react.constantElements']
        }
      }
    };
  }
  return config;
};

exports['default'] = function (args, cb) {
  require('./clean-app')(args);

  console.log('nwb: build-react-app');
  _webpackBuild2['default'](args, buildConfig, cb);
};

module.exports = exports['default'];
