// nwb/lib/commands/serve-react.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _glob = require('glob');

var _glob2 = _interopRequireDefault(_glob);

var _errors = require('../errors');

var _serveReact = require('../serveReact');

var _serveReact2 = _interopRequireDefault(_serveReact);

exports['default'] = function (args, cb) {
  if (args._.length === 1) {
    return cb(new _errors.UserError('usage: nwb serve-react <entry module>'));
  }

  var entry = args._[1];
  if (_glob2['default'].sync(entry).length === 0) {
    return cb(new _errors.UserError('entry module not found: ' + _path2['default'].resolve(entry)));
  }

  console.log('nwb: serve-react');
  _serveReact2['default'](args, {
    entry: entry,
    output: {
      filename: 'app.js',
      path: __dirname,
      publicPath: '/'
    },
    plugins: {
      html: {
        mountId: args['mount-id'] || 'app',
        title: args.title || 'nwb: serve-react'
      }
    }
  }, cb);
};

module.exports = exports['default'];
