// common-accessible-components/lib/components/atoms/ErrorMessage/ErrorMessage.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Alert = require('../Alert');

var _Alert2 = _interopRequireDefault(_Alert);

var _config = require('../../../config');

var _decorators = require('../../../decorators');

var ErrorMessage = function ErrorMessage(props) {
    var children = props.children;

    var other = _objectWithoutProperties(props, ['children']);

    return _react2['default'].createElement(
        _Alert2['default'],
        _extends({
            type: 'error',
            'aria-live': 'assertive',
            screenreaderText: 'Error:'
        }, other),
        props.children
    );
};

ErrorMessage.displayName = 'ErrorMessage';

ErrorMessage.propTypes = {
    className: _react.PropTypes.string,
    children: _react.PropTypes.any.isRequired,
    framework: _react.PropTypes.oneOf(Object.keys(_config.uiFrameworks)).isRequired
};

exports['default'] = ErrorMessage;
module.exports = exports['default'];
